api/README.md

```diff
@@ -0,0 +1,264 @@
+```go
+type ComputeDoubleRequest struct {
+	Input int64
+}
+
+type ComputeDoubleResponse struct {
+	Response int64
+	Overflow bool
+}
+```
+and then the API group's server (`internal/server/dummy/server.go`) needs to define the callbacks to handle requests for all API versions, e.g.:
+```go
+type Server struct{}
+
+func (s *Server) ComputeDouble(ctx context.Context, request *internal.ComputeDoubleRequest, version apiversion.Version) (*internal.ComputeDoubleResponse, error) {
+	in := request.Input64
+	out := 2 * in
+	
+	response := &internal.ComputeDoubleResponse{}
+
+	if sign(in) != sign(out) {
+		// overflow
+		response.Overflow = true
+	} else {
+		response.Response = out
+	}
+
+	return response, nil
+}
+
+func sign(x int64) int {
+	switch {
+	case x > 0:
+		return 1
+	case x < 0:
+		return -1
+	default:
+		return 0
+	}
+}
+```
+All the boilerplate code to:
+ * add a named pipe to the server for each version of the API group, listening for its version's requests, and replying with its version's responses
+ * convert versioned requests to internal representations
+ * and then internal responses back to versioned responses
+ * create clients to talk to that API group & versioned
+is then generated automatically using [gengo](https://github.com/kubernetes/gengo).
+
+The only caveat is that when conversions cannot be made trivially (e.g. when fields from internal and versioned `struct`s have different types), API devs need to define conversion functions. They can do that by creating an (otherwise optional) `internal/server/<api_group_name>/internal/<version>/conversion.go` file, containing functions of the form `func convert_pb_<Type>_To_internal_<Type>(in *pb.<Type>, out *internal.<Type>) error` or `func convert_internal_<Type>_To_pb_<Type>(in *internal.<Type>, out *pb.<Type>) error`; for example, in our `dummy` example above, we need to define a conversion function to account for the different fields in requests and responses from `v1alpha1` to `v1`; so `internal/server/dummy/internal/v1alpha1/conversion.go` could look like:
+```go
+func convert_pb_ComputeDoubleRequest_To_internal_ComputeDoubleRequest(in *pb.ComputeDoubleRequest, out *internal.ComputeDoubleRequest) error {
+	out.Input64 = int64(in.Input32)
+	return nil
+}
+
+func convert_internal_ComputeDoubleResponse_To_pb_ComputeDoubleResponse(in *internal.ComputeDoubleResponse, out *pb.ComputeDoubleResponse) error {
+	i := in.Response
+	if i > math.MaxInt32 || i < math.MinInt64 {
+		// overflow
+		out.Overflow = true
+	} else {
+     		out.Response32 = int32(i)
+    }
+	return nil
+}
+```
+
+## How to change the API
+
+### How to add a new API group
+
+Simply create a new `api/<api_group_name>/<version>/api.proto` file, defining your new service; then generate the Go protobuf code, and run the CSI-proxy generator to generate all the boilerplate code.
+
+FIXME: add more details on which commands to run, and which files to edit when done generating.
```

Comment:
you plan to add this after PR is get approved.

Reply:
Yes indeed. That will depend on the shape of the generator, and I'd rather get this approved before getting too far on writing it.